src/python/pants/core/goals/tailor.py

```diff
@@ -404,16 +419,55 @@ def make_content(bf_path: str, pts: Iterable[PutativeTarget]) -> FileContent:
     return EditedBuildFiles(new_digest, tuple(sorted(created)), tuple(sorted(updated)))
 
 
+def specs_to_dirs(specs: Specs) -> tuple[str, ...]:
+    """Extract cmd-line specs that look like directories.
+
+    Error on all other specs.
+
+    This is a hack that allows us to emulate "directory specs", until we are able to
+    support those more intrinsically.
+
+    TODO: If other goals need "directory specs", move this logic to a rule that produces them.
+    """
```

Comment:
I feel like we could _almost_ unify filesystem and address specs into either files/directories or addresses based on what a consumer wants. So rather than the `Spec` type needing to get more specific, we could define a mapping from globs to addresses and addresses to globs.

We already have this to some degree (informally) in that if we want addresses and get globs on the CLI, we do an owners lookup. The other direction is missing... but I think that that's because it's potentially ambiguous. In this case, the spec that you're treating as a "directory" spec would usually match either a file literal or "the default target" (depending on whether the path is a [file or a directory](https://github.com/pantsbuild/pants/blob/4cb8c4a33b503b47f1b44c9edf23714264283726/src/python/pants/engine/internals/build_files.py#L50-L69)). You might want to use similar code for this match.

But the reason generically converting from `AddressSpecs` to a file match is ambiguous is that when you match the default target, should the only file included be the directory? Or should it be all files owned by the default target?

...all this to say: I don't think that we can generalize "a directory literal" to matching only the directory: as implemented this would need to stay a special case. One way to change the implementation though might be to have this code find all matching files (similar to what `validate` does), and then broaden the match to operate on the containing directories of matched files?

Reply:
Possibly `./pants goal src/python/foo/bar` should mean the same thing as `./pants lint src/python/foo/bar/**/*` ? 
That is kind of what it means in git-land.

Reply:
> Possibly `./pants goal src/python/foo/bar` should mean the same thing as `./pants lint src/python/foo/bar/**/*` ?
> That is kind of what it means in git-land.

Ehhh... maybe. But that would be a very large change. It would essentially be removing the "default target" affordance on the command line, and presumably we want to continue to support it in `BUILD` files.

Reply:
We still have the // disambiguator. Arguably we should deprecate all target address forma that don't use it. That's a pretty massive churn though. That doesn't preclude saying though, that if you want the default target address you must use //.

Reply:
Yeah, I think we need to optimize for new user experience. Existing users will get used to the new thing pretty easily I imagine. I basically only use file specs or address globs now.

Reply:
And in fact I tend to agree with John re `//`.

Reply:
The whole "default target" thing is mildly confusing to users. Especially now that you rarely write dependencies in BUILD files. We could in fact get rid of it with very little inconvenience to the end user, as long as we did something sensible with "directory specs" on the cmd line.

Reply:
I agree generally with biasing towards new users, and I also agree the status quo is confusing. I think we have a path forward to deprecate: https://github.com/pantsbuild/pants/pull/12262#issuecomment-873180141

> Possibly ./pants goal src/python/foo/bar should mean the same thing as ./pants lint src/python/foo/bar/**/* ?

I'm not sure I'm on board with that.

--

@benjyw I'm thinking we have some design work to do on improving specs. I'm a little uncomfortable landing this until we at least agree to a vision for what specs should look like. (Fine to not yet implement). Wdyt?

Reply:
"Design work" seems like big words for actually using something already designed - namely `//`. That makes it totally clear someone is talking about adress specs vs files and directories. I agree though that the rules should be written down.

Reply:
> "Design work" seems like big words for actually using something already designed - namely //.

Yep, but also deciding what `a/b/c` should translate to, e.g. if it's shorthand for `a/b/c/*` vs `a/b/c/**/*`.

Reply:
We can't decide what that means globally right? Pants invokes tools, tools have their own opinion on what a directory means. Most treat it recursively afaict but not all. If we're not tossing out our own principles every other week, Pants should not get in the way of your favorite tools or make their semantics different as a fairly primary goal IIUC.

Reply:
IOW, this is up to each consuming rule to interpret as appropriate.

Reply:
Can we agree that:

A) Files and dirs are natural concepts to the user, targets are not, so filesystem specs (files, dirs and globs of them) should be the norm, address specs should be the exception
B) we can move to requiring `//` as a prefix for address specs, albeit disruptively
C) we can retcon `:` and `::` to mean `*` and `**/*` on filespecs (but more convenient, because the shell doesn't try to interpret them), which would decrease the disruptiveness of B. 
D) The meaning of a directory is context-dependent, but usually means "everything recursively under the directory" because that's how git interprets it, but explicit globs can be used to override that decision if the user wants to be explicit about it. 

?

Reply:
That all sounds right to me.

Reply:
Then I would like to go ahead and merge this PR, unless @Eric-Arellano or @stuhood have further reservations.  I think this is a good feature that users will find useful and intuitive.

Reply:
A-C sound good generally. 

I'm not so sure on D. `a/b/c` would be equivalent to `a/b/c::`, which is the same as `a/b/c/**/*`?

Concretely, I'm thinking about running `./pants test src/python/pants/backend/python`. I would expect it to run all tests in that dir, but not subdirectories. Maybe I think this because I associate Pants with doing the minimum amount of work possible..as in, don't default to running all subdirectory tests unless you added `::` or `**/*` to make it clear.

Reply:
What would pytest do. As much as anything that has to be the right answer. And so, it's verb specific.

Reply:
> We can't decide what that means globally right? Pants invokes tools, tools have their own opinion on what a directory means.

I disagree. Pants aims to provide a consistent interface for the tools you run, such that `./pants lint ::` does the right thing for all N linters you may use. It would be confusing and frustrating if `./pants lint a/b/c` was recursive for Black and Flake8, but not-recursive for Isort and Bandit, for example.

Reply:
I think that was benjy's point D later clauses. If you want recursive add the `::` duffix, if you want just the dir, add the `:` suffix, if you want the default, just specify the dir and let each tool work like you already know it works _if_ you know the underlying tools.

Reply:
I think there are two different user personas here that clash. You're a Pants user and think Pants and Targets (I'd hazard you're in this bucket for the most part), or you're a direct tool user coming to Pants. You're going to do damage to one of these two it seems.

Reply:
> if you want the default, just specify the dir and let each tool work like you already know it works if you know the underlying tools.

I don't think that's desirable from a UX consistency standpoint. 

Independent of UX, this would be tricky to implement..each tool would now need to directly deal with specs, whereas currently `lint.py`, `fmt.py` et al deal with it, such that the individual linter/formatter/test implementations only need code for how to operate on a single file/target.

We went through a lot of effort in https://docs.google.com/document/d/15xphZcFnowytF0Qu2sO1PG7wHAWoyLof7hK0jVa5T8E/edit to ensure specs always behaved consistently when adding FS specs last year, for good reason imo. I'm fine with changing the semantics, but do advocate for the new semantics to be consistent.

Reply:
For me the point is "It would be confusing ..." for who?. UX necessarily picks a `U` or set of `U`s and likely dis-serves some other `U`s. I'd just lobby that we make sure we're really clear about the `U`'s we're thinking of and that we use our expectations as Pants devs as important, but not that important indicators of widers `U`s.

Reply:
I tested this out, and every single linter we use, plus pytest and mypy, treat `src/python/foo` as `src/python/foo/**/*`, except docformatter, which requires files and fails on dirs. 

So I suggest we not worry about hypotheticals here and just go with this. It is clearly enough of a standard that users will accept it if we just apply it uniformly.

Reply:
And it's certainly enough to merge this PR...

Reply:
@benjyw and I discussed more offline. Even though I personally prefer `./pants <goal> a/b/c` to act only on that dir, and not recursively, I agree to disagree.

What I do care about, and @benjyw and I agree with, is that `./pants <goal> a/b/c` behaves consistently, regardless of the goal / tools being run.